xdsl/dialects/experimental/air.py

```diff
@@ -0,0 +1,810 @@
+            src_strides.append(parser.parse_operand())
+            parser.parse_optional_characters(",")
+        parser.parse_characters(")")
+        parser.parse_optional_attr_dict()
+        parser.parse_characters(":")
+        parser.parse_characters("(")
+        parser.parse_type()
+        parser.parse_characters(",")
+        parser.parse_type()
+        parser.parse_characters(")")
+
+        return DmaMemcpyNdOp(
+            None,
+            dst,
+            dst_offsets,
+            dst_sizes,
+            dst_strides,
+            src,
+            src_offsets,
+            src_sizes,
+            src_strides,
+        )
+
+
+@irdl_op_definition
+class ExecuteOp(IRDLOperation):
+    name = "air.execute"
+
+    async_dependencies = var_operand_def(AsyncTokenAttr)
+    async_token = result_def(AsyncTokenAttr)
+    results_ = var_result_def(Attribute)
+    body = region_def()
+
+    traits = traits_def(
+        lambda: frozenset([SingleBlockImplicitTerminator(ExecuteTerminatorOp)])
+    )  # TODO: solve dependency
+
+    def __init__(
+        self,
+        async_dependencies: list[Operation | SSAValue],
+        result_types: list[Attribute] | None,
+        body: Region,
+    ):
+        super().__init__(
+            operands=[async_dependencies],
+            result_types=[AsyncTokenAttr(), result_types],
+            regions=[body],
+        )
+
+    @classmethod
+    def parse(cls, parser: Parser) -> ExecuteOp:
+        async_dependencies: list[Operation | SSAValue] = []
+        if parser.parse_optional_characters("["):
+            while not parser.parse_optional_characters("]"):
+                async_dependencies.append(parser.parse_operand())
+                parser.parse_optional_characters(",")
+
+        result_types: list[Attribute] = []
+        if parser.parse_optional_characters("->"):
+            parser.parse_characters("(")
+            while not parser.parse_optional_characters(")"):
+                result_types.append(parser.parse_type())
+                parser.parse_optional_characters(",")
+
+        body = parser.parse_region()
+
+        return ExecuteOp(async_dependencies, result_types, body)
+
+
+@irdl_op_definition
+class ExecuteTerminatorOp(IRDLOperation):
+    name = "air.execute_terminator"
+
+    results_op = var_operand_def(
+        AnyAttr()
+    )  # even though this is an operand they decided to name it "result" in the original specification
+
+    traits = frozenset([HasParent(ExecuteOp), IsTerminator()])
+
+    def __init__(self, results_op: list[Operation | SSAValue]):
+        super().__init__(operands=[results_op])
+
+    @classmethod
+    def parse(cls, parser: Parser) -> ExecuteTerminatorOp:
+        results_op: list[Operation | SSAValue] = []
+        while not parser.parse_optional_characters(":"):
+            results_op.append(parser.parse_operand())
+        while parser.parse_optional_type():
+            parser.parse_optional_characters(",")
```

Comment:
Also, no `print`?
Personally not against it, just surprised
```suggestion
        parser.parse_optional_undelimited_comma_separated_list(
            parser.parse_type, parser.parse_optional_type
        )
```

Reply:
NB I'm not sure about this one as a suggestion; the method actually make it less readable :thinking: